ssl/quic/quic_rcidm.c

```diff
@@ -0,0 +1,619 @@
+    rcid->type              = type;
+    rcid->state             = RCID_STATE_PENDING;
+
+    if (!ossl_pqueue_RCID_push(rcidm->rcids, rcid, &rcid->pq_idx)) {
+        OPENSSL_free(rcid);
+        return NULL;
+    }
+
+    rcidm_check_rcid(rcidm, rcid);
+    return rcid;
+}
+
+static void rcidm_transition_rcid(QUIC_RCIDM *rcidm, RCID *rcid,
+                                  unsigned int state)
+{
+    unsigned int old_state = rcid->state;
+
+    assert(state >= old_state && state <= RCID_STATE_RETIRING);
+    rcidm_check_rcid(rcidm, rcid);
+    if (state == old_state)
+        return;
+
+    if (rcidm->cur_rcid != NULL && state == RCID_STATE_CUR) {
+        rcidm_transition_rcid(rcidm, rcidm->cur_rcid, RCID_STATE_RETIRING);
+        assert(rcidm->cur_rcid == NULL);
+    }
+
+    if (old_state == RCID_STATE_PENDING) {
+        ossl_pqueue_RCID_remove(rcidm->rcids, rcid->pq_idx);
+        rcid->pq_idx = SIZE_MAX;
+    }
+
+    rcid->state = state;
+
+    if (state == RCID_STATE_CUR) {
+        rcidm->cur_rcid = rcid;
+    } else if (state == RCID_STATE_RETIRING) {
+        if (old_state == RCID_STATE_CUR)
+            rcidm->cur_rcid = NULL;
+
+        ossl_list_retiring_insert_tail(&rcidm->retiring_list, rcid);
+    }
+
+    rcidm_check_rcid(rcidm, rcid);
+}
+
+static void rcidm_free_rcid(QUIC_RCIDM *rcidm, RCID *rcid)
+{
+    if (rcid == NULL)
+        return;
+
+    rcidm_check_rcid(rcidm, rcid);
+
+    switch (rcid->state) {
+    case RCID_STATE_PENDING:
+        ossl_pqueue_RCID_remove(rcidm->rcids, rcid->pq_idx);
+        break;
+    case RCID_STATE_CUR:
+        rcidm->cur_rcid = NULL;
+        break;
+    case RCID_STATE_RETIRING:
+        ossl_list_retiring_remove(&rcidm->retiring_list, rcid);
+        break;
+    default:
+        assert(0);
+        break;
+    }
+
+    OPENSSL_free(rcid);
+}
+
+static void rcidm_handle_retire_prior_to(QUIC_RCIDM *rcidm,
+                                         uint64_t retire_prior_to)
+{
+    RCID *rcid;
+
+    if (retire_prior_to <= rcidm->retire_prior_to)
+        return;
+
+    rcidm->retire_prior_to = retire_prior_to;
+
+    /*
+     * Any RCIDs needing retirement will be at the start of the priority queue,
+     * so just stop once we see a higher sequence number exceeding the
+     * threshold.
+     */
+    while ((rcid = ossl_pqueue_RCID_peek(rcidm->rcids)) != NULL
+           && rcid->seq_num < retire_prior_to)
+        rcidm_transition_rcid(rcidm, rcid, RCID_STATE_RETIRING);
+}
+
+/*
+ * Decision Logic
+ * ==============
+ */
+
+static void rcidm_roll(QUIC_RCIDM *rcidm)
+{
+    RCID *rcid;
+
+    if (ossl_pqueue_RCID_num(rcidm->rcids) < 1)
+        return;
+
+    rcid = ossl_pqueue_RCID_peek(rcidm->rcids);
+    rcidm_transition_rcid(rcidm, rcid, RCID_STATE_CUR);
+
+    ++rcidm->num_changes;
+    rcidm->roll_requested = 0;
+
+    if (rcidm->packets_sent >= PACKETS_PER_RCID)
+        rcidm->packets_sent %= PACKETS_PER_RCID;
+    else
+        rcidm->packets_sent = 0;
```

Comment:
The basic idea here is that if the roll was triggered by the packet threshold, we should do the roll consistently at the correct number of packets and thus handle this accurately in case we already transmitted more than the threshold number of packets. Whereas for other roll triggers, the cadence is necessarily reset.